docs/cheat-sheets/python-code-injection.mdx

```diff
@@ -0,0 +1,332 @@
+
+- Ensure evaluated content is not definable by external sources.
+- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-code-run.dangerous-interactive-code-run" />
+
+### 1.E. Using **subinterpreter** to run code
+
+`_xxsubinterpreters.run_string` is an internal Python function that interprets the string as Python code.
+If unverified user data can reach `run_string`, this is a command injection vulnerability.
+A malicious actor can inject a malicious string to execute arbitrary Python code.
+
+Example:
+
+```python
+import _xxsubinterpreters
+
+# Value supplied by user
+user_input = "print('pwned')"
+
+# Vulnerable
+_xxsubinterpreters.run_string(_xxsubinterpreters.create(), user_input)
+```
+
+#### References
+
+- [**subinterpreters** documentation](https://pythondev.readthedocs.io/subinterpreters.html)
+
+#### Mitigation
+
+Do not let user input in `_xxsubinterpreters` methods. Alternatively:
+
+- Ensure evaluated content is not definable by external sources.
+- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-subinterpreters-run-string.dangerous-subinterpreters-run-string" />
+
+### 1.F. Running **subinterpreter** from regression tests package
+
+`run_in_subinterp` is a function from regression tests package for Python that runs code in subinterpreter.
+This is dangerous if external data can reach this function call because it allows a malicious actor to run arbitrary Python code.
+
+Example:
+
+```python
+import _testcapi
+
+# Value supplied by user
+user_input = "print('pwned')"
+
+# Vulnerable
+_testcapi.run_in_subinterp(user_input)
+
+
+from test import support
+
+# Value supplied by user
+user_input = "print('pwned')"
+
+# Vulnerable
+support.run_in_subinterp(user_input)
+```
+
+#### References
+
+- [**test** module documentation](https://docs.python.org/3/library/test.html)
+
+#### Mitigation
+
+Do not let user input in `run_in_subinterp` function. Alternatively:
+
+- Ensure evaluated content is not definable by external sources.
+- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-testcapi-run-in-subinterp" />
+
+## 2. Abusing built-in functions
+
+### 2.A. Accessing dictionary with current global/local symbol table
+
+`globals()` and `locals()` return a dictionary representing the current global/local symbol table. Using non-static data to retrieve values from this table is extremely dangerous because it may allow an attacker to execute arbitrary code on the system. 
```

Comment:
```suggestion
`globals()` and `locals()` return a dictionary representing the current global or local symbol table. Using non-static data to retrieve values from this table is extremely dangerous because it can allow an attacker to execute arbitrary code on the system. 
```